src/ucp/core/ucp_worker.c

```diff
@@ -771,8 +908,33 @@ static ucs_status_t ucp_worker_add_resource_ifaces(ucp_worker_h worker)
             iface_params.mode.device.dev_name = resource->tl_rsc.dev_name;
         }
 
+        status = ucp_worker_iface_open(worker, tl_id, &iface_params,
+                                       &worker->ifaces[iface_id++]);
+        if (status != UCS_OK) {
+            return status;
+        }
+    }
+
+    if (!ctx_tl_bitmap) {
+        /* Context bitmap is not set, need to select the best tl resources */
+        tl_bitmap = 0;
+        status    = ucp_worker_select_best_ifaces(worker, &tl_bitmap);
+        if (status != UCS_OK) {
+            return status;
+        }
+        ucs_assert(tl_bitmap);
+
+        /* Cache tl_bitmap on the context, so the next workers would not need
+         * to select best ifaces. */
+        ucs_atomic_cswap64(&context->tl_bitmap, 0ul, tl_bitmap);
```

Comment:
not sure we need atomic, if several threads write they would write same value anyway, no?